# Sperm control cohort mosaicism
This repository collects pipelines, codes, and some intermediate results for the study of mosaic SNV/Indels for sperm, blood, and other samples of a control cohort. Raw data of this study is available on SRA under [PRJNA660493](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA660493/) and [PRJNA588332](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA588332/).

### 1. Pipelines for the process of whole-genome sequencing data
#### 1.1 Pipelines for WGS data process and quality control

#### 1.2 Pipelines for mosaic SNV/indel calling and variant annotations


### 2. Pipelines for the process of Targeted Amplicon Sequencing (TAS)
#### 2.1 Pipelines for TAS data alignment and processing
#### 2.2 Pipelines for AF quantification and variant annotations

### 3. Pipelines for the data analysis, variant filtering, comprehensive annotations, and statistical analysis
#### 3.1 Pipelines for mosaic variant determination, annotations, and plotting
#### 3.2 Pipelines for statistically analysis, QC, and the related plotting
#### 3.3 Codes for statistical modeling
[Codes](https://github.com/shishenyxx/Sperm_control_cohort_mosaicism/blob/master/Mutation_accumulation_model/fit_afs.py) for the estimation of accumulation of mutations through a stepwise exponential regression regression model.
